File: 396-rotate-function/Sony/396. Rotate Function.py
class Solution(object):
    def maxRotateFunction(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        a_len = len(A)

        weighted_sum = 0
        total = 0
        max_value  = 0
        for idx, value in enumerate(A):
            weighted_sum = weighted_sum + idx * value
            total = total + value

        max_value = weighted_sum
        pivot = a_len - 1

        while pivot > 0:
            weighted_sum = weighted_sum + total - a_len * A[pivot]
            pivot = pivot - 1
            max_value = max(max_value, weighted_sum)

        return max_value

if __name__ == '__main__':
    sol = Solution()

    A = [4, 3, 2, 6]


    print sol.maxRotateFunction(A)

File: 396-rotate-function/leochang/396-rotate-function.cpp
class Solution {
public:
	int maxRotateFunction(vector<int>& A) {
		int size = A.size();
		long long sum = 0;
		long long weightedSum = 0;

		if (size == 0) return 0;

		for (int i = 0; i < size; ++i){
			sum += A[i];
			weightedSum += A[i] * i;
		}

		long long maxSum = weightedSum;

		for (int i = size -1; i > 0; --i) {
			weightedSum = weightedSum + sum - size * A[i];
			maxSum = max(maxSum, weightedSum);
		}

		return maxSum;
	}
};
